package/piksi_system_daemon/src/async-child.c

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (C) 2017 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#include <unistd.h>
+#include <czmq.h>
+
+struct async_child_ctx {
+  zloop_t *loop;
+  zmq_pollitem_t pollitem;
+  int pid;
+  int exit_status;
+
+  void (*output_callback)(const char *buf, void *external_context);
+  void (*exit_callback)(int status, void *external_context);
+  void *external_context;
+
+  struct async_child_ctx *next;
+};
+
+static struct async_child_ctx *async_children;
+
+/* This pipe is used to synchronise reaping of the dead child from SIGCHLD
+ * with the zmq event loop. */
+static int wakepipe[2];
+static zmq_pollitem_t wakeitem;
+
+/* We use this unbuffered fgets function alternative so that the czmq loop
+ * will keep calling us with output until we've read it all.
+ */
+static int raw_fgets(char *str, size_t len, int fd)
+{
+  size_t i;
+  len--;
+  for (i = 0; i < len; i++) {
+    int r = read(fd, &str[i], 1);
+    if ((r < 0) && (errno != EAGAIN))
+      return r;
+    if ((r <= 0) || (str[i] == '\n'))
+      break;
+  }
+  str[i++] = 0;
+  return i > 1 ? i : 0;
+}
+
+static int async_output_cb(zloop_t *loop, zmq_pollitem_t *item, void *arg)
+{
+  struct async_child_ctx *ctx = arg;
+  char buf[128];
+
+  if (raw_fgets(buf, sizeof(buf), item->fd) > 0) {
+    if (ctx->output_callback)
+      ctx->output_callback(buf, ctx->external_context);
+  }
+
+  return 0;
+}
+
+static int async_cleanup_dead_child(zloop_t *loop, zmq_pollitem_t *item, void *arg)
+{
+  struct async_child_ctx *ctx;
+  read(item->fd, &ctx, sizeof(ctx));
+
+  /* Child is dead, notify observer and clean up */
+  ctx->exit_callback(ctx->exit_status, ctx->external_context);
+
+  if (ctx == async_children) {
+    async_children = ctx->next;
+  } else {
+    for (struct async_child_ctx *c = async_children; c; c = c->next) {
+      if (c->next == ctx) {
+        c->next = ctx->next;
+        break;
+      }
+    }
+  }
+  free(ctx);
+
+  return 0;
+}
+
+int async_spawn(zloop_t *loop, char **argv,
```

Comment:
Is this `argv` desirable, or is it better to take a single string command line and run with `sh -c`?